aggregator/handler/forward.go

```diff
@@ -0,0 +1,231 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package handler
+
+import (
+	"errors"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3aggregator/aggregator"
+	"github.com/m3db/m3metrics/protocol/msgpack"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+)
+
+const (
+	tcpProtocol = "tcp"
+)
+
+var (
+	errEmptyServerList = errors.New("empty server list")
+	errHandlerClosed   = errors.New("handler is closed")
+	errBufferQueueFull = errors.New("buffer queue is full")
+)
+
+type dropType int
+
+const (
+	dropCurrent dropType = iota
+	dropOldestInQueue
+)
+
+type dialWithTimeoutFn func(protocol string, addr string, timeout time.Duration) (net.Conn, error)
+type writeToConnFn func(conn *net.TCPConn, data []byte) (int, error)
+
+type forwardHandler struct {
+	sync.RWMutex
+
+	servers             []string
+	log                 xlog.Logger
+	connectTimeout      time.Duration
+	connectionKeepAlive bool
+	reconnectRetrier    xretry.Retrier
+
+	bufCh             chan msgpack.Buffer
+	dialWithTimeoutFn dialWithTimeoutFn
+	writeToConnFn     writeToConnFn
+	wg                sync.WaitGroup
+	closed            bool
+}
+
+// NewForwardHandler creates a new forwarding handler.
+// TODO(xichen): add metrics.
+func NewForwardHandler(
+	servers []string,
+	opts ForwardHandlerOptions,
+) (aggregator.Handler, error) {
+	numServers := len(servers)
+	if numServers == 0 {
+		return nil, errEmptyServerList
+	}
+
+	h := &forwardHandler{
+		servers:             servers,
+		log:                 opts.InstrumentOptions().Logger(),
+		connectTimeout:      opts.ConnectTimeout(),
+		connectionKeepAlive: opts.ConnectionKeepAlive(),
+		reconnectRetrier:    opts.ReconnectRetrier(),
+		bufCh:               make(chan msgpack.Buffer, opts.QueueSize()),
+		dialWithTimeoutFn:   net.DialTimeout,
+		writeToConnFn:       writeToConn,
+	}
+
+	// Initiate the connections.
+	h.wg.Add(numServers)
+	for _, addr := range servers {
+		conn, err := h.tryConnect(addr)
+		if err != nil {
+			h.log.WithFields(
+				xlog.NewLogField("address", addr),
+				xlog.NewLogErrField(err),
+			).Error("error connecting to server")
+		}
+		go h.forwardToConn(addr, conn)
```

Comment:
Yeah so this is so that when the servers start up, we already have most if not all connections set up and can start draining queues right away. Otherwise you might run into a situation where the servers start taking in writes, but we are still in the middle of trying to establish connections.